flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1177,144 +1197,124 @@ def _infer_volumes_for_test(node_uuid, node_hostname, compute_instance_id,
+        if dataset_id in node_state.devices:
+            datasets[dataset_id] = DiscoveredDataset(
+                state=DatasetStates.ATTACHED,
+                dataset_id=dataset_id,
+                maximum_size=dataset.maximum_size or 0,
+                device_path=node_state.devices[dataset_id],
+                blockdevice_id=_create_blockdevice_id_for_test(dataset_id),
+            )
+        else:
+            datasets[dataset_id] = DiscoveredDataset(
+                state=DatasetStates.NON_MANIFEST,
+                dataset_id=dataset_id,
+                maximum_size=dataset.maximum_size or 0,
+                blockdevice_id=_create_blockdevice_id_for_test(dataset_id),
+            )
 
-    def shared_state_changes(self):
-        return self.node_state, self.nonmanifest_datasets
+    if node_state.manifestations is not None:
+        for dataset_id, manifestation in node_state.manifestations.items():
+            dataset_id = UUID(dataset_id)
+            datasets[dataset_id] = DiscoveredDataset(
+                state=DatasetStates.MOUNTED,
+                dataset_id=dataset_id,
+                maximum_size=manifestation.dataset.maximum_size or 0,
+                device_path=node_state.devices[dataset_id],
+                blockdevice_id=_create_blockdevice_id_for_test(dataset_id),
+                mount_point=node_state.paths[unicode(dataset_id)],
+            )
 
+    return BlockDeviceDeployerLocalState(
+        hostname=node_state.hostname,
+        node_uuid=node_state.uuid,
+        datasets=datasets,
+        volumes=volumes,
+    )
 
-class FakeBlockDeviceDeployerLocalStateTests(SynchronousTestCase):
-    """
-    Tests for ``BlockDeviceDeployerLocalState``.
-    """
-    def setUp(self):
-        self.node_uuid = uuid4()
-        self.hostname = u"192.0.2.1"
 
-    def test_provides_ilocalstate(self):
-        """
-        Verify that ``BlockDeviceDeployerLocalState`` instances provide the
-        ILocalState interface.
-        """
-        local_state = FakeBlockDeviceDeployerLocalState(
-            node_state=EMPTY_NODE_STATE,
-            nonmanifest_datasets=NonManifestDatasets(datasets={}),
-            volumes=[],
-        )
-        self.assertTrue(
-            verifyObject(ILocalState, local_state)
-        )
-
-    def test_shared_changes(self):
-        """
-        ``shared_state_changes`` returns a ``NodeState`` with
-        the ``node_uuid`` and ``hostname`` from the
-        BlockDeviceDeployerLocalState`` and a
-        ``NonManifestDatasets``.
-        """
-        local_state = FakeBlockDeviceDeployerLocalState(
-            node_state=NodeState(
-                uuid=self.node_uuid,
-                hostname=self.hostname,
-                manifestations={},
-                paths={},
-                devices={},
-                applications=None
+DISCOVERED_DATASET_STRATEGY = tagged_union_strategy(
+    DiscoveredDataset,
+    {
+        'dataset_id': uuids(),
```

Comment:
`tagged_union_strategy` is a strategy that will generate things from all possible states. You just need to pass the possible attributes, and it will generate possibilities from there.